symengine/logic.cpp

```diff
@@ -146,4 +146,103 @@ int Piecewise::compare(const Basic &o) const
     RCP<const Piecewise> t = o.rcp_from_this_cast<Piecewise>();
     return unified_compare(get_vec(), t->get_vec());
 }
+
+And::And(const set_boolean &s) : args{s}
+{
+    SYMENGINE_ASSERT(is_canonical(s));
+}
+
+std::size_t And::__hash__() const
+{
+    std::size_t seed = AND;
+    for (const auto &a : args)
+        hash_combine<Basic>(seed, *a);
+    return seed;
+}
+
+vec_basic And::get_args() const
+{
+    vec_basic v(args.begin(), args.end());
+    return v;
+}
+
+bool And::__eq__(const Basic &o) const
+{
+    return is_a<And>(o) and unified_eq(args, static_cast<const And &>(o).args);
+}
+
+int And::compare(const Basic &o) const
+{
+    SYMENGINE_ASSERT(is_a<And>(o))
+    return unified_compare(args, static_cast<const And &>(o).args);
+}
+
+bool And::is_canonical(const set_boolean &container)
+{
+    return container.size() >= 2;
+}
+
+Or::Or(const set_boolean &s) : args{s}
+{
+    SYMENGINE_ASSERT(is_canonical(s));
+}
+
+std::size_t Or::__hash__() const
+{
+    std::size_t seed = OR;
+    for (const auto &a : args)
+        hash_combine<Basic>(seed, *a);
+    return seed;
+}
+
+vec_basic Or::get_args() const
+{
+    vec_basic v(args.begin(), args.end());
+    return v;
+}
+
+bool Or::__eq__(const Basic &o) const
+{
+    return is_a<Or>(o) and unified_eq(args, static_cast<const Or &>(o).args);
+}
+
+int Or::compare(const Basic &o) const
+{
+    SYMENGINE_ASSERT(is_a<Or>(o))
+    return unified_compare(args, static_cast<const Or &>(o).args);
+}
+
+bool Or::is_canonical(const set_boolean &container)
+{
+    return container.size() >= 2;
```

Comment:
You also need to check that the none of the elements are `BooleanAtom`s. Basically, `is_canonical` should do all the checks that you do when you are creating the object. In this case, `is_canonical` should check all the cases that `and_or` does